### DISCUSSION TITLE
Avoid double GA page views with query on SSG url

### DISCUSSION THREAD

**Discussion by valse** [vercel/next.js#12306] [Help] [answered]
Hi, I've a SSG site and I've implemented the `react-ga` support like in your example, adding the listener to the `routeChangeComplete` event on the `_app` page.

I added a check to the `isFallback` status too so I don't count this page views but I can't exclude the double page render when I set a query string, for example in the `search?q=test` page... This is because the query is populated in the second render.
Any suggestions? Thanks

**Comment by oMatej** [comment, marked as answer]
You can consider doing something like this:

```
import React from 'react'
import App from 'next/app'
import Router from 'next/router'
import { initGA, logPageView } from '../utils/analytics'

export default class MyApp extends App {
  componentDidMount() {
    const { router } = this.props
    const { asPath, query } = router || {}

    initGA()

    if (!asPath.includes('?') || (asPath.includes('?') && !isEmpty(query))) {
      logPageView()
    }

    Router.events.on('routeChangeComplete', logPageView)
  }

  render() {
    const { Component, pageProps } = this.props
    return <Component {...pageProps} />
  }
}

**Reply by valse** [reply]
What is the `isEmpty` function?

**Reply by oMatej** [reply]
@valse  ah, sorry, in this case it was just `lodash/isEmpty`. But you can use anything to check if `query` object is empty.

Since you are interested in SSG, then it might be enough for you to check only `!asPath.includes('?')` condition.

What I noticed with `next.js` is following:
- client-side only pages during initial render contains proper `asPath`, but `query` is empty;
- server-side pages during initial render contains proper `asPath` and `query` is filled with query parameters;
- client-side only pages when contains any query parameter, will invoke `routeChangeComplete` right after render;

So to avoid these duplicated `pageview` events, in some conditions we should not call `pageview` on component mount, and do it only on `routeChangeComplete`.

**Reply by valse** [reply]
> Since you are interested in SSG, then it might be enough for you to check only `!asPath.includes('?')` condition.

Yes, I confirm; with this setting on page refresh with a query string only the pageView of the  `routeChangeComplete` event is fired and the first one skipped... in the other cases the opposite!
